subcommand/connect-init/command.go

```diff
@@ -81,19 +92,84 @@ func (c *Command) Run(args []string) int {
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Hit maximum retries for consul login: %s", err))
+			return 1
+		}
+		c.UI.Info("Consul login complete")
+	}
+	if c.flagSkipServiceRegistrationPolling {
+		return 0
+	}
 
+	// Now wait for the service to be registered. Do this by querying the Agent for a service
+	// which maps to this one.
+	// In the case of ACLs this will match the serviceAccountName, we query on this.
+	// If ACLs are disabled we query all services and search through
+	// the list for a service with `meta["pod-name"]` that matches this pod.
+	data := ""
 	err = backoff.Retry(func() error {
-		err := common.ConsulLogin(c.consulClient, c.flagBearerTokenFile, c.flagACLAuthMethod, c.flagTokenSinkFile, c.flagMeta)
-		if err != nil {
-			c.UI.Error(fmt.Sprintf("Consul login failed; retrying: %s", err))
+		if c.flagACLAuthMethod == "" {
+			filter := fmt.Sprintf("Kind != `%s` and Meta.pod-name == %s and Meta.k8s-namespace == %s", "connect-proxy", c.flagPodName, c.flagPodNamespace)
+			serviceList, err := c.consulClient.Agent().ServicesWithFilter(filter)
+			if err != nil {
+				c.UI.Error(fmt.Sprintf("Unable to get agent service: %s", err))
+				return err
+			}
+			for _, y := range serviceList {
+				// TODO: in theory we've already filtered enough.. can we just return?
+				if y.Kind != "connect-proxy" && y.Meta["pod-name"] == c.flagPodName && y.Meta["k8s-namespace"] == c.flagPodNamespace {
+					c.UI.Info(fmt.Sprintf("Registered pod has been detected: %s", y.Meta["pod-name"]))
+					data = fmt.Sprintf("%s-%s-%s", c.flagPodName, y.ID, "sidecar-proxy")
+					return nil
+				}
+			}
+			return fmt.Errorf("Unable to find registered service")
+		} else {
+			// If ACLs are enabled we don't have permission to go through the list of all services
+			svc, _, err := c.consulClient.Agent().Service(c.flagServiceAccountName, &api.QueryOptions{})
+			if err != nil {
+				c.UI.Error(fmt.Sprintf("Unable to write proxyid out: %s", err))
+				return err
+			} else {
+				if svc == nil {
+					c.UI.Info(fmt.Sprintf("unable to fetch registered service for %v", c.flagServiceAccountName))
+					return fmt.Errorf("Unable to find registered service")
+				}
+				data = fmt.Sprintf("%s-%s-%s", c.flagPodName, svc.ID, "sidecar-proxy")
+			}
+			return nil
 		}
-		return err
-	}, backoff.WithMaxRetries(backoff.NewConstantBackOff(1*time.Second), numLoginRetries))
+	}, backoff.WithMaxRetries(backoff.NewConstantBackOff(1*time.Second), serviceRegistrationPollingRetries))
+	if err != nil {
+		c.UI.Error("Timed out waiting for service registration")
+		return 1
+	}
+	// Write the proxyid to the shared volume.
+	err = ioutil.WriteFile(c.flagProxyIDFile, []byte(data), 0444)
 	if err != nil {
-		c.UI.Error(fmt.Sprintf("Hit maximum retries for consul login: %s", err))
+		c.UI.Error(fmt.Sprintf("Unable to write proxyid out: %s", err))
 		return 1
 	}
-	c.UI.Info("Consul login complete")
+	c.UI.Info("Bootstrapping completed")
```

Comment:
```suggestion
	c.UI.Info("Service registration completed")
```
Not sure what this should say, if anything. "bootstrapping" is definitely not complete yet because Envoy hasn't been bootstrapped. Thoughts?